src/main/java/org/broadinstitute/barclay/argparser/TaggedArgumentParser.java

```diff
@@ -0,0 +1,333 @@
+package org.broadinstitute.barclay.argparser;
+
+import org.apache.commons.lang3.tuple.Pair;
+import org.broadinstitute.barclay.utils.Utils;
+
+import java.util.*;
+
+/**
+ * The parser used by the {@link CommandLineArgumentParser} for handling tagged argument strings. Tagged arguments
+ * contain a tag name and optional attribute/value pairs, and are of the form:
+ *
+ *     --argument_name:logical_name argument_value
+ *
+ * or:
+ *
+ *     --argument_name:logical_name,key1=value1,key2=value2 argument_value
+ *
+ * The logical name is required; key/value pairs are optional. @Argument annotated fields that require tagged
+ * values must implement {@link TaggedArgument}, and the command line parser will require tag names for any such
+ * fields.
+ *
+ * The parsing occurs in two phases:
+ *
+ * 1) Phase 1 occurs before the option parser is presented with the arguments. In phase 1, the option name is
+ * peeled off from the raw option string (which includes that tag string), and the resulting tag string and accompanying
+ * raw argument value are stored as a pair in a hash map for later retrieval via a key. The key is a string that is
+ * constructed by concatenating the original raw option string (including the actual short/long hyphen prefix, option
+ * name, and tag string), and the raw argument value together to ensure the key is unique:
+ *
+ *      Key: --argument_name:logical_name,key1=value1,key2=value2:argument_value
+ *
+ * The pair object stored in the map contains only the (previously peeled off) tag string, and the raw argument value that
+ * was provided by the user. The command line parser replaces the original arguments provided by the user with the option
+ * name and the key:
+ *
+ *          User provides:              "--argument_name:logical_name,key1=value1,key2=value2" "raw_argument_value"
+ *          Parser is presented with:   "--argument_name" "--argument_name:logical_name,key1=value1,key2=value2:argument_value"
```

Comment:
Just checked in the debugger - its definitely not there. I also just verified that you can't use a shortName from one arg as the longName for another. So I think its fine. Its the comments that need to change.